src/Management/src/Endpoint/SpringBootAdminClient/SpringBootAdminPeriodicRefresh.cs

```diff
@@ -0,0 +1,107 @@
+        SpringBootAdminRefreshRunner runner, ILogger<SpringBootAdminPeriodicRefresh> logger)
+    {
+        ArgumentNullException.ThrowIfNull(optionsMonitor);
+        ArgumentNullException.ThrowIfNull(timeProvider);
+        ArgumentNullException.ThrowIfNull(runner);
+        ArgumentNullException.ThrowIfNull(logger);
+
+        TimeSpan safeInterval = InfiniteWhenZero(optionsMonitor.CurrentValue.RefreshInterval);
+
+        _runner = runner;
+        _logger = logger;
+        _periodicTimer = new PeriodicTimer(safeInterval, timeProvider);
+        _timerTask = TimerLoopAsync(safeInterval);
+        _optionsChangeToken = optionsMonitor.OnChange(options => ChangeInterval(options.RefreshInterval));
+    }
+
+    private async Task TimerLoopAsync(TimeSpan interval)
+    {
+        try
+        {
+            _logger.LogDebug("Starting periodic refresh loop with interval {Interval}.", interval);
+
+            do
+            {
+                _logger.LogDebug("Starting refresh cycle.");
+
+                try
+                {
+                    await _runner.RunAsync(_timerTokenSource.Token);
+                }
+                catch (Exception exception) when (!exception.IsCancellation())
+                {
+                    _logger.LogWarning(exception, "Refresh cycle failed.");
+                }
+            }
+            while (await _periodicTimer.WaitForNextTickAsync(_timerTokenSource.Token));
+        }
+        catch (OperationCanceledException)
+        {
+#pragma warning disable S6667 // Logging in a catch clause should pass the caught exception as a parameter.
+            // Justification: The exception contains no useful information. Logging it suggests something crashed, while this is expected behavior.
+            _logger.LogDebug("Stopped periodic refresh loop.");
+#pragma warning restore S6667 // Logging in a catch clause should pass the caught exception as a parameter.
+        }
+    }
+
+    private void ChangeInterval(TimeSpan interval)
+    {
+        TimeSpan safeInterval = InfiniteWhenZero(interval);
+
+        if (safeInterval != _periodicTimer.Period)
+        {
+            _periodicTimer.Period = safeInterval;
+            _logger.LogDebug("Refresh interval changed to {Interval}.", safeInterval);
+        }
+    }
+
+    private static TimeSpan InfiniteWhenZero(TimeSpan interval)
+    {
+        // Translate disabled to a very high value, so we can resume when configuration changes.
+        return interval <= TimeSpan.Zero ? Timeout.InfiniteTimeSpan : interval;
+    }
+
+    public async Task StopAsync(CancellationToken cancellationToken)
+    {
+        _logger.LogDebug("Signaling to stop periodic refresh loop.");
+        await DisposeAsync();
+
+        _logger.LogDebug("Starting cleanup.");
+        await _runner.CleanupAsync(cancellationToken);
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        if (!_isDisposed)
+        {
+            _isDisposed = true;
```

Comment:
Nothing jumps out at me as likely to break here, but shouldn't `_isDisposed` be set _after_ everything is disposed

Reply:
The problem with that is when an exception is thrown half-way, we don't know which line caused it. So the first fields may already have been disposed, while the last lines haven't executed yet. Running dispose again would try to re-run code, potentially resulting in an exception because the object is already disposed. For example, the call to `await _timerTokenSource.CancelAsync()` throws `ObjectDisposedException`. So in that case, there's no way to recover. By setting the flag in advance, a second call to Dispose won't throw. The misery remains, but it's a little bit better because calling dispose multiple times shouldn't throw.

The guidelines are:
- Calling Dispose multiple times should not throw.
- Do not call Dispose multiple times.